API Client Generation: Add documentation (#102056)
grafana/grafana: pkg/tests/apis/openapi_snapshots/API_CLIENT_GENERATION.md

> This folder contains a rendered OpenAPI file for each group/version. The “real” OpenAPI is generated by the running server, but the files here are used to build static frontend clients.

Author: I copied this part from the old README.mdx doc that was already there, but there is no issue on removing these lines 😄

<br/>

### 3. Create the API definition
In the folder `../public/app/features/{your_group_name}/api/` you have to create the `baseAPI.ts` file for your group. This file should have the following content:

```jsx
import { createApi } from '@reduxjs/toolkit/query/react';

import { createBaseQuery } from 'app/api/createBaseQuery';
import { getAPIBaseURL } from 'app/api/utils';

export const BASE_URL = getAPIBaseURL('dashboard.grafana.app', 'v0alpha1');

export const baseAPI = createApi({
  reducerPath: 'dashboardAPI',
  baseQuery: createBaseQuery({
    baseURL: BASE_URL,
  }),
  endpoints: () => ({}),
});
```
This is the API definition for the specific group you're working with, where `getAPIBaseURL` should have the proper `group` and `version` as parameters. The `reducePath` should also be modified to match `group + API`: `dashboard` will be `dashboardAPI`, `iam` will be `iamAPI` and so on. 

### 4. Add the output information
Open [generate-rtk-apis.ts](https://github.com/grafana/grafana/blob/main/scripts/generate-rtk-apis.ts) and add the following information:

> | Data | Descritpion |

Author: We should tighten the list of needed properties to just the ones that need to be modified. After reducing the list, I'd add a description with the purpose of that property. Even the official documentation does not have a lot of information 🙃

Reviewer: the ones we definitely need are: `apiFile`, `schemaFile`, `apiImport`, `filterEndpoints`, `tag` and `hooks` (both these need to be `true`)

Reviewer: `argSuffix`/`responseSuffix` we should probably not mention, and i'll try to remove it from anywhere that's using it in a followup PR

Author: > the ones we definitely need are: apiFile, schemaFile, apiImport, filterEndpoints, tag and hooks (both these need to be true)

Can you tell me what they do and why they must be set to true? I think it is a valuable info for the doc

Reviewer: `hooks` is so it generates and exports the hooks
`tags` is to automatically attach tags to endpoints. this is needed for proper cache invalidation. see [here](https://redux-toolkit.js.org/rtk-query/usage/automated-refetching#:~:text=RTK%20Query%20uses,an%20active%20subscription.) in the docs

Reviewer: actually i've just realised, we probably want `hooks: false` 😂 

we only want to export the hooks from the index file, not the generated one

| apiFile | File with the group's API definition. |
| schemaFile | File with the schema that was automatically created in the second step. Although it is in openapi_snapshots, you should link the one saved in `data/openapi`.|
|apiImport| Function name exported in the API definition (baseAPI.ts file).|
| filterEndpoints | The `operationId` of the particular route you want to work with. You can check the available operationIds in the specific group's spec file. As seen in the `migrate-to-cloud` one, it is an array|
| argSuffix | |
| responseSuffix | |
| tag | |
| hooks | |

> More info in [Redux Toolkit](https://redux-toolkit.js.org/rtk-query/usage/code-generation#simple-usage) 

In our example, the information added will be:

```jsx
'../public/app/features/dashboard/api/endpoints.gen.ts': {
    apiFile: '../public/app/features/dashboard/api/baseAPI.ts',
    schemaFile: '../data/openapi/dashboard.grafana.app-v0alpha1.json',
    apiImport: 'baseAPI',
    filterEndpoints: ['createDashboard'],
    argSuffix: 'Arg',
    responseSuffix: 'Response',
    tag: true,
    hooks: true,
},
```

> ### 5. Run the API Client script

Reviewer: ```suggestion
### 5. Run the API client generation script
```

Then, we are ready to run the script to create the API client:

```jsx
yarn generate-apis
```

This will create an `endpoints.gen.ts` file in the path specified in the previous step.

### 6. Create the index file for your hooks

In the same `api` folder where the `endpoints.gen.ts` file has been saved, you have to create an index file from which you can import the types and hooks needed. By doing this, we selectively export hooks/types from `endpoints.gen.ts`.

In our case, the dashboard index will be like:

```jsx
import { generatedAPI } from './endpoints.gen';

export const dashboardAPI = generatedAPI;
export const { useCreateDashboardMutation} = dashboardAPI;
// eslint-disable-next-line no-barrel-files/no-barrel-files
export { type Dashboard } from './endpoints.gen';

```

> There are some use cases where the hook will not work, and that is a clue to see if it needs to be modified. The hooks can be tweaked by using `enhanceEndpoints`.

Reviewer: ```suggestion
There are some use cases where the generated hook might not work out of the box, and that is a clue to see if it needs to be modified. The hooks can be tweaked by using `enhanceEndpoints`.
```

TODO: Add examples of `enhanceEndpoints` usage

> ### 7. Add information to the store

Reviewer: ```suggestion
### 7. Add reducers and middleware to the redux store
```

> The last, but not least, step to be done is adding the middleware and reducers to the store. You have available the oficial documentation in [RTK Query](https://redux-toolkit.js.org/tutorials/rtk-query#add-the-service-to-your-store)

Reviewer: maybe move the docs link to the end, as it's kinda supplementary info

In Grafana, the reducers are added to `public/app/core/reducers/root.ts`:

> ```jsx
>   const rootReducers = {
>     ...sharedReducers,
>     ...alertingReducers,
>     ...teamsReducers,
>     ...apiKeysReducers,
>     ...foldersReducers,
>     ...dashboardReducers,
>     ...exploreReducers,
>     ...dataSourcesReducers,
>     ...usersReducers,
>     ...serviceAccountsReducer,
>     ...userReducers,
>     ...invitesReducers,
>     ...organizationReducers,
>     ...browseDashboardsReducers,
>     ...ldapReducers,
>     ...importDashboardReducers,
>     ...panelEditorReducers,
>     ...panelsReducers,
>     ...templatingReducers,
>     ...supportBundlesReducer,
>     ...authConfigReducers,
>     plugins: pluginsReducer,
>     [alertingApi.reducerPath]: alertingApi.reducer,
>     [publicDashboardApi.reducerPath]: publicDashboardApi.reducer,
>     [browseDashboardsAPI.reducerPath]: browseDashboardsAPI.reducer,
>     [cloudMigrationAPI.reducerPath]: cloudMigrationAPI.reducer,
>     [iamApi.reducerPath]: iamApi.reducer,
>     [userPreferencesAPI.reducerPath]: userPreferencesAPI.reducer,
>     [provisioningAPI.reducerPath]: provisioningAPI.reducer,
>     [folderAPI.reducerPath]: folderAPI.reducer,
>     [dashboardAPI.reducerPath]: dashboardAPI.reducer,
>   };
> ```

Reviewer: i'd just really slim these examples down to only show the necessary info, e.g.

```suggestion
```jsx
  import { dashboardAPI } from '<pathToYourAPI>';
  const rootReducers = {
    ...,
    [dashboardAPI.reducerPath]: dashboardAPI.reducer,
  };
```
```

> And the middlewares are added to `public/app/store/configureStore.ts`:

Reviewer: can (relative) link to the file here

> ```jsx
> export function configureStore(initialState?: Partial<StoreState>) {
>   const store = reduxConfigureStore({
>     reducer: createRootReducer(),
>     middleware: (getDefaultMiddleware) =>
>       getDefaultMiddleware({ thunk: true, serializableCheck: false, immutableCheck: false }).concat(
>         listenerMiddleware.middleware,
>         alertingApi.middleware,
>         publicDashboardApi.middleware,
>         browseDashboardsAPI.middleware,
>         cloudMigrationAPI.middleware,
>         userPreferencesAPI.middleware,
>         iamApi.middleware,
>         provisioningAPI.middleware,
>         folderAPI.middleware,
>         dashboardAPI.middleware
>         ...extraMiddleware
>       ),
>     devTools: process.env.NODE_ENV !== 'production',
>     preloadedState: {
>       navIndex: buildInitialState(),
>       ...initialState,
>     },
>   });
> ```

Reviewer: same here about slimming down

```suggestion
```jsx
import { dashboardAPI } from '<pathToYourAPI>';
export function configureStore(initialState?: Partial<StoreState>) {
  const store = reduxConfigureStore({
    reducer: createRootReducer(),
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({ thunk: true, serializableCheck: false, immutableCheck: false }).concat(
        ...,
        dashboardAPI.middleware
      ),
    ...,
  });
```
```
